[Docs][RISCV] Document RISC-V vector codegen

lukel97 opened the pull request:
This is a revival of https://reviews.llvm.org/D142348, and attempts to document how RVV semantics can be expressed in LLVM IR as well as how codegen works in the backend.

Parts of this are taken from the original RFC https://lists.llvm.org/pipermail/llvm-dev/2020-October/145850.html, but I've largely rewritten this from the original differential revision to exclude explaining the specification itself and instead just focus on the LLVM specific bits. (I figured that there's better material available elsewhere for learning about RVV itself)

I've also updated it to include as much as I know about fixed vector codegen as well as the recent changes to vsetvli insertion. Let me know if I'm missing anything else that would be useful to document.

lukel97 committed:
[Docs][RISCV] Document RISC-V vector codegen

This is a revival of https://reviews.llvm.org/D142348, and attempts to document how RVV semantics can be expressed in LLVM IR as well as how codegen works in the backend.

Parts of this are taken from the original RFC https://lists.llvm.org/pipermail/llvm-dev/2020-October/145850.html, but I've largely rewritten this from the original differential revision to exclude explaining the specification itself and instead just focus on the LLVM specific bits. (I figured that there's better material available elsewhere for learning about RVV itself)

I've also updated it to include as much as I know about fixed vector codegen as well as the recent changes to vsetvli insertion. Let me know if I'm missing anything else that would be useful to document.

wangpc-pp reviewed (commented):
Great! I don't see any problem here!
I think we need some unfamiliar eyes on this doc, because we may miss some important details since we all are familiar with RVV implementation.

wangpc-pp commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,285 @@
+
+   Unlike RISC-V intrinsics, VP intrinsics are target agnostic so they can be emitted from other optimisation passes in the middle-end (like the loop vectorizer). They also support fixed length vector types.
+
+SelectionDAG lowering
```
It may be worthwhile to mention what the `mask`/`vl` will be for these three lowering paths.

lukel97 replied:
I've added a table that hopefully summarizes this

lukel97 committed:
Add a missing word, reference langref for scalable vector types

wangpc-pp reviewed (commented):
I have written an internel doc about this before, and this is how I organized the content:
![图片](https://github.com/llvm/llvm-project/assets/137158460/ff0a548f-1dc1-45ca-abd2-311b087100ec)
Hopefully this can help us to perfect this doc. :-)

tschuett commented:
Thanks! It is inconsistent whether whiches deserve a comma. A glossary at the bottom would be nice. What is AVL?

tschuett commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,285 @@
+Scalable vector types are of the form ``<vscale x n x ty>``, which indicate a vector with a multiple of ``n`` elements of type ``ty``. ``n`` and ``ty`` then end up controlling LMUL and SEW respectively.
+
+LLVM supports only ``ELEN=32`` or ``ELEN=64``, so ``vscale`` is defined as ``VLEN/64`` (see ``RISCV::RVVBitsPerBlock``).
+This makes the LLVM IR types stable between the two ``ELEN`` s considered, i.e. every LLVM IR scalable vector type has exactly one corresponding pair of element type and LMUL, and vice-versa.
```
A comma after `, i.e., `

tschuett commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,285 @@
+The only legal vector MVTs on RISC-V are scalable, so fixed length vectors need to be custom lowered and performed in a scalable container type:
+
+1. The fixed length vector operands are inserted into scalable containers via ``insert_subvector``. The container size is chosen to have a minimum size big enough to fit the fixed length vector (see ``getContainerForFixedLengthVector``).
+2. The operation is then performed via a scalable **VL (vector length) node**. These are custom nodes that contain an AVL operand which is set to the size of the fixed length vector, and are defined in RISCVInstrInfoVVLPatterns.td.
```
`AVL operand, which`

tschuett commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,285 @@
+Instruction selection
+=====================
+
+VL and VTYPE need to be configured correctly, so we can't just directly select the underlying vector MachineInstrs. Instead a layer of pseudo instructions get selected which carry the extra information needed to emit the necessary ``vsetvli`` instructions later.
```
which without a comma.

dtcxzyw commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,285 @@
+
+   // custom lowered to:
+
+   t18: nxv2i32 = RISCVISD::ADD_VL t2, t4, undef:nxv2i32, t6, Constant:i64<8>
```
Can you explain the meaning of `undef:nxv2i32` here?

lukel97 replied:
Yeah good idea, that's the passthru operand which just happens to be undef for this fixed vector lowering.

As an aside it's sometimes called the merge operand, but I think we should eventually standardise this to the term passthru to avoid confusion with vmerge.vv. But that's for another day

rofirrim commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,285 @@
+Mask patterns
+-------------
+
+For the VL patterns we only match to masked pseudos to reduce the size of the match table, even if the node's mask is all ones and could be an unmasked pseudo. The ``RISCVDAGToDAGISel::doPeepholeMaskedRVV`` will detects that the mask is all ones during post-processing and convert it into its unmasked form.
```
typo: will detects → will detect

lukel97 commented:
> Thanks! It is inconsistent whether whiches deserve a comma. A glossary at the bottom would be nice. What is AVL?

The AVL is the application vector length, i.e. the operand to vsetvli. I think it's quite easily confused with VL, which is the actual register. Those _VL nodes should probably be called _AVL nodes, and likewise everywhere where we have a "VL" value. The VL register isn't guaranteed to end up as that value, it's just what RISCVInsertVSETVLI will use the vsetvli's AVL.

rofirrim commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,285 @@
+
+   .. code-block:: llvm
+
+       %c = call @llvm.riscv.vadd.nxv4i32.nxv4i32(
```
I think this intrinsic is the same as above and it is not a masked variant.

rofirrim commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,285 @@
+
+   t5: nxv4i32 = add t2, t4
+
+RISC-V vector intrinsics are also always scalable and so don't need custom lowering:
```
Did you mean "are also always legal" or maybe "are also always legal for scalable types"?

lukel97 replied:
I meant to point out the fact that they only operate on scalable types, and thus are always legal. But I agree the current wording is confusing, will clarify

rofirrim commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,285 @@
+
+VL nodes often have a passthru or mask operand, which are usually set to undef and all ones for fixed length vectors.
+
+The ``insert_subvector`` and ``extract_subvector`` nodes responsible for wrapping and unwrapping will get combined away, and eventually we will lower all fixed vector types to scalable. Note that the vectors at the interface of a function are always scalable vectors.
```
Just to confirm I understand the last statement: do you mean fixed vectors at the interface of a function would get somehow scalarised or will also be contained in the corresponding scalable vector type?

lukel97 replied:
The latter, I will clarify that. cc @wangpc-pp did you mention anything about fixed vectors in the PEI section of your internal doc?

wangpc-pp replied:
> The latter, I will clarify that. cc @wangpc-pp did you mention anything about fixed vectors in the PEI section of your internal doc?

Not much, I just wrote down things about RVV objects on stack and spilling/restore.

lukel97 committed:
Address some review comments

- Fix use of which
- Clarify what VP operands are used for in VL nodes
- Clarify why RVV intrinsics are legal
- Be more precise on usage of "legal" fixed vector types. They are legal, there's just no patterns for them
- Clarify how fixed length vector arguments are passed in

Still some comments to address!

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+
+RVV adds 32 ``VLEN`` sized registers, where ``VLEN`` is an unknown constant to the compiler. To be able to represent ``VLEN`` sized values, the RISC-V backend takes the same approach as AArch64's SVE and uses `scalable vector types <https://llvm.org/docs/LangRef.html#t-vector>`_.
+
+Scalable vector types are of the form ``<vscale x n x ty>``, which indicate a vector with a multiple of ``n`` elements of type ``ty``. ``n`` and ``ty`` then end up controlling LMUL and SEW respectively.
```
I'd suggest "which indicates" since we're referring to "the form", not "vector types". Reads a little better to me.

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+- <vscale x 32 x i1>
+- <vscale x 64 x i1>
+
+Two types with the same ratio SEW/LMUL will have the same related mask type. For instance, two different comparisons one under SEW=64, LMUL=2 and the other under SEW=32, LMUL=1 will both generate a mask <vscale x 2 x i1>.
```
"SEW/LMUL ratio"

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+              <vscale x 4 x i32> %passthru,
+	      <vscale x 4 x i32> %a,
+	      <vscale x 4 x i32> %b,
+	      i64 %avl
```
should there not be a `<vscale x 4 x i1> %mask` argument here as well? And an extra `i64` (is that the mask/tail policy? I'm a bit out of the loop)?

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+
+   The only valid types are scalable vector types.
+
+3. :doc:`Vector predication (VP) intrinsics </Proposals/VectorPredication>`
```
Could also link to https://llvm.org/docs/LangRef.html#vector-predication-intrinsics, I think.

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+
+   // custom lowered to:
+
+       t2: nxv2i32,ch = CopyFromReg t0, Register:nxv2i32 %0
```
I've never fully understood how DAGs are indented, but when comparing two of them it's a little confusing to have the indentation of a node change all of a sudden. I wonder if it's worth custom-formatting these examples to keep them as stable as possible?

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+     t16: nxv2i32 = RISCVISD::ADD_VL t2, t6, undef:nxv2i32, t15, Constant:i64<4>
+   t17: v4i32 = extract_subvector t16, Constant:i64<0>
+
+VL nodes often have a passthru or mask operand, which are usually set to undef and all ones for fixed length vectors.
```
back-ticked ``undef``?

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+
+.. note::
+
+   The only ``insert_subvector`` and ``extract_subvector`` nodes that make it through lowering are those that can be performed as an exact subregister insert or extract. This means that any fixed length vector ``insert_subvector`` and ``extract_subvector`` nodes that aren't legalized must lie on a register group boundary, so the exact ``VLEN`` must be known at compile time (i.e. compiled with ``- ...
```
there's an `i.e.` here which could do with a comma, for consistency.

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+
+   t18: nxv2i32 = RISCVISD::ADD_VL t2, t4, undef:nxv2i32, t6, Constant:i64<8>
+
+The VP EVL and mask are used for the VL node's AVL and mask respectively, whilst the passthru is set to undef. A passthru can be emulated to get tail/mask undisturbed behaviour by using ``@llvm.vp.merge``. It will get lowered as a ``vmerge``, but will likely be merged back into the underlying instruction's mask via ``RISCVDAGToDAGISel::performCombineVMergeAndVOps``.
```
back-ticked `undef` here too, if desired.

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+Each vector instruction has multiple pseudo instructions defined in ``RISCVInstrInfoVPseudos.td``.
+
+The pseudos have operands for the AVL and SEW (encoded as a power of 2), as well as potentially the mask, policy or rounding mode if applicable.
+The passhthru operand is tied to the destination register to control the inactive/tail elements.
```
typo: `passhthru` -> `passthru`.

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+
+For each possible LMUL there is a variant of the pseudo instruction, as it affects the register class needed for the operands, and similarly there are ``_MASK`` variants that control whether or not the instruction is masked.
+
+For scalable vectors that should use VLMAX, the AVL is set to a sentinel value of -1.
```
back-tick `-1`, I think?

4vtomat commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+
+   .. code-block:: llvm
+
+       %c = call @llvm.riscv.vadd.mask.nxv4i32.nxv4i32(
```
Does it miss the mask argument?

4vtomat replied:
And policy argument~

frasercrmck replied:
ach, beat me to it!

4vtomat replied:
Actually, you beat me to it according your comment time lol~

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+The pseudos have operands for the AVL and SEW (encoded as a power of 2), as well as potentially the mask, policy or rounding mode if applicable.
+The passhthru operand is tied to the destination register to control the inactive/tail elements.
+
+For each possible LMUL there is a variant of the pseudo instruction, as it affects the register class needed for the operands, and similarly there are ``_MASK`` variants that control whether or not the instruction is masked.
```
I think maybe here, or elsewhere, is a good place to provide an example of the referenced instructions in MachineIR form (what does one per LMUL look like, what does _MASK look like, what are the operands, and in what order?)

I hope that doesn't become too unwieldy but I think it could be useful for people trying to see the entire flow.

lukel97 replied:
Good idea, it might be a bit unwieldy but it's showcases the fact that the pseudos are unwieldy themselves

topperc commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+Scalable vector types are of the form ``<vscale x n x ty>``, which indicate a vector with a multiple of ``n`` elements of type ``ty``. ``n`` and ``ty`` then end up controlling LMUL and SEW respectively.
+
+LLVM supports only ``ELEN=32`` or ``ELEN=64``, so ``vscale`` is defined as ``VLEN/64`` (see ``RISCV::RVVBitsPerBlock``).
+This makes the LLVM IR types stable between the two ``ELEN`` s considered, i.e., every LLVM IR scalable vector type has exactly one corresponding pair of element type and LMUL, and vice-versa.
```
This is a bug not a feature. Because of it we can't support VLEN=32.

topperc commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+       %c = call @llvm.riscv.vadd.mask.nxv4i32.nxv4i32(
+              <vscale x 4 x i32> %passthru,
+	      <vscale x 4 x i32> %a,
+	      <vscale x 4 x i32> %b,
```
This is missing a mask operand

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+RISCVInsertVSETVLI
+==================
+
+After vector registers are allocated, the RISCVInsertVSETVLI pass will insert the necessary vsetvlis for the pseudos.
```
I think in general, when referencing code concepts like pass names `RISCVInsertVSETVLI` or specific ISA instructions (`vsetvli`), back-ticking/escaping them is best.

lukel97 replied:
Agreed: I noticed that the specification doesn't back-tick the design-time constants like EVL/AVL though, so I'll try and unback-tick those.

topperc commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+	      i64 %avl
+	    )
+
+   Both allow setting the AVL as well as controlling the inactive/tail elements via the passthru operand, but the masked variant also provides operands for the mask and ``vta``/``vma`` policy bits.
```
The example masked intrinsic above doesn't have a policy operand.

frasercrmck commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+Pseudo expansion and printing
+=============================
+
+After scalar register allocation, the ``RISCVExpandPseudoInsts.cpp`` pass expands out the ``PseudoVSETVLI``.
```
maybe "expands the `PseudoVSETVLI` instructions" - "expands out" doesn't sound right in this context.

topperc commented on llvm/docs/RISCV/RISCVVectorExtension.rst:
```diff
@@ -0,0 +1,289 @@
+
+   t5: nxv4i32 = add t2, t4
+
+This is because the TableGen patterns for RVV are only defined for scalable vector types.
```
I'm not sure how this sentence adds to the previous statement. The use of "only" makes it seems like its trying to contrast with fixed vectors, but that's not spelled out very well.

lukel97 replied:
Agreed, I've deferred trying to explain this to the fixed-length vector subsection.

frasercrmck reviewed (commented):
In general this looks great, thank you!

I'm not arguing particularly strongly in favour of this suggestion -- and despite me being a native speaker I'm still not 100% on this -- but when the phrase "fixed length" is used as an adjective, as in "fixed-length vector", then it's technically best to use the hyphenated form. As in, "a fixed-length vector has a fixed length". Same with certain uses of "target-agnostic" vs "target agnostic". Again, I don't want to be too prescriptive on this matter and it's not a blocker for me.

lukel97 committed:
Fix masked intrinsic example

lukel97 commented:
> I'm not arguing particularly strongly in favour of this suggestion -- and despite me being a native speaker I'm still not 100% on this -- but when the phrase "fixed length" is used as an adjective, as in "fixed-length vector", then it's technically best to use the hyphenated form. As in, "a fixed-length vector has a fixed length". Same with certain uses of "target-agnostic" vs "target agnostic". Again, I don't want to be too prescriptive on this matter and it's not a blocker for me.

Thanks for the review! Agreed, that seems to be what LangRef.rst uses. I think I was too busy trying to not write "fixed vector"

lukel97 committed:
Add original RFC to see also

lukel97 committed:
Clarify that defining vscale = VLEN / 64 prevents VLEN = 32 from being supported

lukel97 committed:
indicate->indicates, reword how n and ty control LMUL and SEW

lukel97 committed:
Address review comments

- Put more code terms inside backticks
- Update VP link
- fixed length -> fixed-length
- Add table comparing operands and properties of different LLVM IR constructs
- Add missing mask intrinsic operands
- Add table listing different pseudo permutations
- Add missing policy operand to PseudoVADD examples
- Use MIR in masked pattern examples
- i.e. -> i.e.,
- Move vp vmerge passthru emulation from SelectionDAG section to LLVM IR section
- Fix typos
- Update SelectionDAG indentation

lukel97 committed:
Don't backtick design time constants, backtick registers. Correct VL -> AVL

lukel97 commented:
Ping

topperc reviewed (approved):
LGTM

wangpc-pp reviewed (approved):
LGTM! Let's take the first step!

lukel97 committed:
Wording tweaks, VTYPE -> ``vtype``, remove redundant phrasing
